File: php/src/detectors/HeaderInjectionDetector.php
<?php

namespace FireWTWall\Detectors;

class HeaderInjectionDetector
{
    /** CRLF characters used for HTTP response splitting */
    private const CRLF_PATTERN = '/[\r\n]|%0[aAdD]|\\\\r|\\\\n/i';

    /**
     * Scan request headers for CRLF injection and host-header injection.
     *
     * @param  array  $headers  Normalised header array (lowercase keys)
     * @param  string $host     Value of the Host header
     * @return array{rule:string,severity:string,matched:string,source:string}|null
     */
    public static function scan(array $headers, string $host): ?array
    {
        // 1. CRLF injection in any header
        foreach ($headers as $name => $value) {
            if (preg_match(self::CRLF_PATTERN, $value, $m)) {
                return [
                    'rule'     => 'crlf-injection',
                    'severity' => 'critical',
                    'matched'  => substr($m[0], 0, 120),
                    'source'   => 'header:' . $name,
                ];
            }
        }

        // 2. Host header injection
        if ($host !== '' && preg_match('/[\/?\r\n@#]/', $host, $m)) {
            return [
                'rule'     => 'host-header-injection',
                'severity' => 'high',
                'matched'  => substr($m[0], 0, 120),
                'source'   => 'header:host',
            ];
        }

        return null;
    }
}

File: aspnet/src/detectors/HeaderInjectionDetector.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace FireWTWall.Detectors
{
    /// <summary>
    /// Detects CRLF injection and Host header injection.
    /// </summary>
    public static class HeaderInjectionDetector
    {
        private static readonly Regex CrlfPattern = new Regex(
            @"[\r\n]|%0[aAdD]|\\r|\\n",
            RegexOptions.IgnoreCase | RegexOptions.Compiled);

        private static readonly Regex HostInjectionPattern = new Regex(
            @"[/?\r\n@#]",
            RegexOptions.Compiled);

        public static DetectorResult Scan(Dictionary<string, string> headers, string host)
        {
            // 1. CRLF injection in any header
            foreach (var kv in headers)
            {
                Match m = CrlfPattern.Match(kv.Value);
                if (m.Success)
                    return new DetectorResult("crlf-injection", "critical",
                        m.Value.Length > 120 ? m.Value.Substring(0, 120) : m.Value,
                        "header:" + kv.Key);
            }

            // 2. Host header injection
            if (!string.IsNullOrEmpty(host))
            {
                Match m = HostInjectionPattern.Match(host);
                if (m.Success)
                    return new DetectorResult("host-header-injection", "high",
                        m.Value.Length > 120 ? m.Value.Substring(0, 120) : m.Value,
                        "header:host");
            }

            return null;
        }
    }
}
